relay-system/src/service.rs

```diff
@@ -426,8 +502,105 @@ pub trait Service: Sized {
 
     /// Starts the service in the current runtime and returns an address for it.
     fn start(self) -> Addr<Self::Interface> {
-        let (addr, rx) = channel();
+        let (addr, rx) = channel(Self::name());
         self.spawn_handler(rx);
         addr
     }
+
+    /// Returns a unique name for this service implementation.
+    ///
+    /// This is used for internal diagnostics and uses the fully qualified type name of the service
+    /// implementor by default.
+    fn name() -> &'static str {
+        std::any::type_name::<Self>()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    struct MockMessage;
+
+    impl Interface for MockMessage {}
+
+    impl FromMessage<Self> for MockMessage {
+        type Response = NoResponse;
+
+        fn from_message(message: Self, _: ()) -> Self {
+            message
+        }
+    }
+
+    struct MockService;
+
+    impl Service for MockService {
+        type Interface = MockMessage;
+
+        fn spawn_handler(self, mut rx: Receiver<Self::Interface>) {
+            tokio::spawn(async move {
+                while rx.recv().await.is_some() {
+                    tokio::time::sleep(BACKLOG_INTERVAL * 2).await;
+                }
+            });
+        }
+
+        fn name() -> &'static str {
+            "mock"
+        }
+    }
+
+    #[test]
+    fn test_backpressure_metrics() {
+        let rt = tokio::runtime::Builder::new_current_thread()
+            .enable_time()
+            .build()
+            .unwrap();
+
+        let _guard = rt.enter();
+        tokio::time::pause();
```

Comment:
How does this work? Is the time still paused after calling `tokio::time::sleep`? That is, calling `sleep(X)` increases the paused `now` time by `X`?

Reply:
The relevant information is in the ["Auto-advance" section](https://docs.rs/tokio/latest/tokio/time/fn.pause.html#auto-advance) on the docs of pause: The runtime skips ahead to the next waiting timer, in our case a sleep. We can think about it like skipping idle time. Also, the timer does not advance in between, so multiple calls to `Instant::now()` will always give the same instant back.

There is also `tokio::time::advance()`, but that does not run all paused timers, which we want in this case.